internal/deploy/application/remote.go

```diff
@@ -0,0 +1,288 @@
+/*
+ *  *******************************************************************************
+ *  * Copyright (c) 2019 Edgeworx, Inc.
+ *  *
+ *  * This program and the accompanying materials are made available under the
+ *  * terms of the Eclipse Public License v. 2.0 which is available at
+ *  * http://www.eclipse.org/legal/epl-2.0
+ *  *
+ *  * SPDX-License-Identifier: EPL-2.0
+ *  *******************************************************************************
+ *
+ */
+
+package deployapplication
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/eclipse-iofog/iofogctl/internal/config"
+	"github.com/eclipse-iofog/iofogctl/pkg/iofog/client"
+	"github.com/eclipse-iofog/iofogctl/pkg/util"
+)
+
+type remoteExecutor struct {
+	namespace          string
+	opt                *config.Application
+	microserviceByName map[string]*config.Microservice
+	client             *client.Client
+	agentsByName       map[string]*client.AgentInfo
+	catalogByID        map[int]*client.CatalogItemInfo
+	catalogByName      map[string]*client.CatalogItemInfo
+}
+
+func microserviceArrayToMap(a []config.Microservice) (result map[string]*config.Microservice) {
+	result = make(map[string]*config.Microservice)
+	for i := 0; i < len(a); i++ {
+		result[a[i].Name] = &a[i]
+	}
+	return
+}
+
+func newRemoteExecutor(namespace string, opt *config.Application) *remoteExecutor {
+	exe := &remoteExecutor{
+		namespace:          namespace,
+		opt:                opt,
+		microserviceByName: microserviceArrayToMap(opt.Microservices),
+	}
+
+	return exe
+}
+
+//
+// Deploy application using remote controller
+//
+func (exe *remoteExecutor) Execute() (err error) {
+	// Get Controllers from namespace
+	controllers, err := config.GetControllers(exe.namespace)
+
+	// Do we actually have any controllers?
+	if err != nil {
+		util.PrintError("You must deploy a Controller to a namespace before deploying any Agents")
+		return
+	}
+
+	// Did we have more than one controller?
+	if len(controllers) != 1 {
+		err = util.NewInternalError("Only support 1 controller per namespace")
+		return
+	}
+
+	// Init remote resources
+	if err = exe.init(&controllers[0]); err != nil {
+		return
+	}
+
+	// Validate application definition (routes, agents, etc.)
+	if err = exe.validate(); err != nil {
+		return
+	}
+
+	// Deploy application
+	if err = exe.deploy(); err != nil {
+		return
+	}
+	return nil
+}
+
+func (exe *remoteExecutor) init(controller *config.Controller) (err error) {
+	exe.client = client.New(controller.Endpoint)
+	if err = exe.client.Login(client.LoginRequest{Email: controller.IofogUser.Email, Password: controller.IofogUser.Password}); err != nil {
+		return
+	}
+	listAgents, err := exe.client.ListAgents()
+	if err != nil {
+		return
+	}
+	exe.agentsByName = make(map[string]*client.AgentInfo)
```

Comment:
Could you re-use your other function which converts array to map? Assuming you can make it generic by using interface. If so, you can move it to util pkg if need

Reply:
Without golang generics, there is too many small variations to make a generic array to map function